ledger/simulation/resources.go

```diff
@@ -605,3 +607,696 @@ func (p *resourcePolicy) AvailableBox(app basics.AppIndex, name string, operatio
+	return PopulatedResourceArrays{
+		Accounts: slices.Concat(r.prefilledAccounts, r.accounts),
+		Assets:   slices.Concat(r.prefilledAssets, r.assets),
+		Apps:     slices.Concat(r.prefilledApps, r.apps),
+		Boxes:    slices.Concat(r.prefilledBoxes, r.boxes),
+	}
+}
+
+// resourcePopulator is used to populate app resources for a transaction group
+type resourcePopulator struct {
+	txnResources   map[int]*txnResources
+	appCallIndexes []int
+	groupSize      int
+}
+
+func (p *resourcePopulator) addTransaction(txn transactions.Transaction, groupIndex int, consensusParams config.ConsensusParams) {
+	p.txnResources[groupIndex] = &txnResources{
+		prefilledAssets:    []basics.AssetIndex{},
+		prefilledApps:      []basics.AppIndex{},
+		prefilledAccounts:  []basics.Address{},
+		prefilledBoxes:     []logic.BoxRef{},
+		accountsFromFields: make(map[basics.Address]struct{}),
+
+		assets:       []basics.AssetIndex{},
+		apps:         []basics.AppIndex{},
+		accounts:     []basics.Address{},
+		boxes:        []logic.BoxRef{},
+		maxTotalRefs: consensusParams.MaxAppTotalTxnReferences,
+		maxAccounts:  consensusParams.MaxAppTxnAccounts,
+		maxBoxes:     consensusParams.MaxAppBoxReferences,
+		maxApps:      consensusParams.MaxAppTxnForeignApps,
+		maxAssets:    consensusParams.MaxAppTxnForeignAssets,
+	}
+
+	// The Sender will always be implicitly available for every transaction type
+	p.txnResources[groupIndex].addAddressFromField(txn.Sender)
+
+	switch txn.Type {
+	case protocol.ApplicationCallTx:
+		p.txnResources[groupIndex].prefilledAssets = append(p.txnResources[groupIndex].prefilledAssets, txn.ForeignAssets...)
+
+		p.txnResources[groupIndex].prefilledApps = append(p.txnResources[groupIndex].prefilledApps, txn.ForeignApps...)
+
+		p.txnResources[groupIndex].prefilledAccounts = append(p.txnResources[groupIndex].prefilledAccounts, txn.Accounts...)
+
+		for _, box := range txn.Boxes {
+			ref := logic.BoxRef{App: txn.ForeignApps[box.Index], Name: string(box.Name)}
+			p.txnResources[groupIndex].prefilledBoxes = append(p.txnResources[groupIndex].prefilledBoxes, ref)
+		}
+
+		p.txnResources[groupIndex].appFromField = txn.ApplicationID
+
+	case protocol.AssetTransferTx:
+		p.txnResources[groupIndex].assetFromField = txn.XferAsset
+		p.txnResources[groupIndex].addAddressFromField(txn.AssetReceiver)
+		p.txnResources[groupIndex].addAddressFromField(txn.AssetCloseTo)
+		p.txnResources[groupIndex].addAddressFromField(txn.AssetSender)
+
+	case protocol.PaymentTx:
+		p.txnResources[groupIndex].addAddressFromField(txn.Receiver)
+		p.txnResources[groupIndex].addAddressFromField(txn.CloseRemainderTo)
+
+	case protocol.AssetConfigTx:
+		p.txnResources[groupIndex].assetFromField = txn.ConfigAsset
+
+	case protocol.AssetFreezeTx:
+		p.txnResources[groupIndex].assetFromField = txn.FreezeAsset
+		p.txnResources[groupIndex].addAddressFromField(txn.FreezeAccount)
+	}
+}
+
+func (p *resourcePopulator) addAccount(addr basics.Address) error {
+	var err error
```

Comment:
Nit: move`err` declaration down to just before the second for loop. Same in `addAsset`, `addApp`, etc.